src/pytorch_lightning/utilities/data.py

```diff
@@ -322,12 +322,44 @@ def _dataloader_init_kwargs_resolve_sampler(
     batch_sampler = getattr(dataloader, "batch_sampler")
     is_predicting = mode == RunningStage.PREDICTING
     # checking the batch sampler type is different than PyTorch default.
-    if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
-        batch_sampler = type(batch_sampler)(
-            sampler,
-            batch_size=batch_sampler.batch_size,
-            drop_last=(False if is_predicting else batch_sampler.drop_last),
-        )
+    batch_cls = type(batch_sampler)
+    if batch_sampler is not None and (batch_cls is not BatchSampler or is_predicting):
+        if hasattr(batch_sampler, "__pl_saved_args"):
+            args = list(batch_sampler.__pl_saved_args)
+            kwargs = batch_sampler.__pl_saved_kwargs
+            arg_names = batch_sampler.__pl_saved_arg_names
+
+            if is_predicting:
+                success, args, kwargs = _inject_arg_to_saved(args, kwargs, arg_names, "drop_last", False)
+                if not success:
+                    rank_zero_warn(
+                        "Trying to inject `drop_last=False` into batch sampler since you are predicting, however it "
+                        f"seems the class `{batch_cls}` does not support it. Your predictions might be incomplete. "
+                        "To mitigate this, expose `drop_last` in the `__init__` method of your custom class."
+                    )
+
+            success, args, kwargs = _inject_arg_to_saved(args, kwargs, arg_names, "sampler", sampler)
+            if not success:
+                raise MisconfigurationException(
+                    "Trying to inject modified sampler into batch sampler, however it seems the class "
+                    f"`{batch_cls}` does not support argument called sampler. To mitigate this, "
+                    "expose argument `sampler` in the `__init__` method of your custom class."
+                )
+
+            batch_sampler = batch_cls(*args, **kwargs)
+        else:
+            try:
+                batch_sampler = batch_cls(
+                    sampler,
+                    batch_size=batch_sampler.batch_size,
+                    drop_last=(False if is_predicting else batch_sampler.drop_last),
+                )
+            except TypeError:
```

Comment:
This should check that the TypeError is a true positive as done in 

https://github.com/Lightning-AI/lightning/blob/daf7cec01e526c7503b99bfbf9e4c2048b8c8514/src/pytorch_lightning/utilities/data.py#L193-L202

Reply:
Done